ipns/record.go

```diff
@@ -3,124 +3,466 @@ package ipns
 import (
 	"bytes"
 	"errors"
-
-	pb "github.com/ipfs/boxo/ipns/pb"
+	"fmt"
+	"sort"
+	"time"
 
 	"github.com/gogo/protobuf/proto"
+	ipns_pb "github.com/ipfs/boxo/ipns/pb"
+	"github.com/ipfs/boxo/path"
+	"github.com/ipfs/boxo/util"
+	"github.com/ipfs/go-cid"
 	logging "github.com/ipfs/go-log/v2"
-	record "github.com/libp2p/go-libp2p-record"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/codec/dagcbor"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	basicnode "github.com/ipld/go-ipld-prime/node/basic"
 	ic "github.com/libp2p/go-libp2p/core/crypto"
 	"github.com/libp2p/go-libp2p/core/peer"
-	pstore "github.com/libp2p/go-libp2p/core/peerstore"
+	"github.com/multiformats/go-multibase"
+	"github.com/multiformats/go-multihash"
+	"go.uber.org/multierr"
 )
 
 var log = logging.Logger("ipns")
 
-var _ record.Validator = Validator{}
+type ValidityType int64
 
-// RecordKey returns the libp2p record key for a given peer ID.
-func RecordKey(pid peer.ID) string {
-	return "/ipns/" + string(pid)
-}
+// ValidityEOL means "this record is valid until {Validity}". This is currently
+// the only supported Validity type.
+const ValidityEOL ValidityType = 0
 
-// Validator is an IPNS record validator that satisfies the libp2p record
-// validator interface.
-type Validator struct {
-	// KeyBook, if non-nil, will be used to lookup keys for validating IPNS
-	// records.
-	KeyBook pstore.KeyBook
+// Record represents an [IPNS Record].
+//
+// [IPNS Record]: https://specs.ipfs.tech/ipns/ipns-record/
+type Record struct {
+	pb   ipns_pb.IpnsEntry
+	node datamodel.Node
 }
```

Comment:
The way I implemented the `Record` abstraction keeps the Protobuf in memory. Therefore, `NewRecord` takes the option `CompatibleWithV1`. I decided to do this because it makes it much easier for testing.

However, it could be interesting to represent `Record` and only make the decision of compatibility with V1 when marshaling. This is still up for debate. It will make testing harder, but still possible.

```suggestion
// Record represents an [IPNS Record].
//
// [IPNS Record]: https://specs.ipfs.tech/ipns/ipns-record/
type Record struct {
	// For validation of unmarshalled records.
	signatureV1 []byte
	signatureV2 []byte
	
	pk   ic.PubKey
	node datamodel.Node
}
```